File: android/app/src/main/java/com/smilepile/data/repository/PhotoRepository.kt
package com.smilepile.data.repository

import com.smilepile.data.models.Photo
import kotlinx.coroutines.flow.Flow

interface PhotoRepository {
    suspend fun insertPhoto(photo: Photo): Long
    suspend fun insertPhotos(photos: List<Photo>)
    suspend fun updatePhoto(photo: Photo)
    suspend fun deletePhoto(photo: Photo)
    suspend fun deletePhotoById(photoId: Long)
    suspend fun getPhotoById(photoId: Long): Photo?
    suspend fun getPhotoByPath(path: String): Photo?
    suspend fun getPhotosByCategory(categoryId: Long): List<Photo>
    fun getPhotosByCategoryFlow(categoryId: Long): Flow<List<Photo>>
    suspend fun getPhotosInCategories(categoryIds: List<Long>): List<Photo>
    fun getPhotosInCategoriesFlow(categoryIds: List<Long>): Flow<List<Photo>>
    suspend fun getAllPhotos(): List<Photo>
    fun getAllPhotosFlow(): Flow<List<Photo>>
    suspend fun deletePhotosByCategory(categoryId: Long)
    suspend fun getPhotoCount(): Int
    suspend fun getPhotoCategoryCount(categoryId: Long): Int

    // Remove from library (app only, not device)
    suspend fun removeFromLibrary(photo: Photo)
    suspend fun removeFromLibraryById(photoId: Long)

    // Search and filter methods
    fun searchPhotos(searchQuery: String): Flow<List<Photo>>
    fun searchPhotosInCategory(searchQuery: String, categoryId: Long): Flow<List<Photo>>
    fun getPhotosByDateRange(startDate: Long, endDate: Long): Flow<List<Photo>>
    fun getPhotosByDateRangeAndCategory(startDate: Long, endDate: Long, categoryId: Long): Flow<List<Photo>>
    fun searchPhotosWithFilters(
        searchQuery: String,
        startDate: Long,
        endDate: Long,
        favoritesOnly: Boolean?,
        categoryId: Long?
    ): Flow<List<Photo>>
}

File: ios/SmilePile/Data/Repositories/PhotoRepository.swift
import Foundation
import Combine

public protocol PhotoRepository {
    // CRUD Operations
    func insertPhoto(_ photo: Photo) async throws -> Int64
    func insertPhotos(_ photos: [Photo]) async throws
    func updatePhoto(_ photo: Photo) async throws
    func deletePhoto(_ photo: Photo) async throws
    func deletePhotoById(_ photoId: Int64) async throws

    // Retrieval Operations
    func getPhotoById(_ photoId: Int64) async throws -> Photo?
    func getPhotoByPath(_ path: String) async throws -> Photo?
    func getAllPhotos() async throws -> [Photo]
    func getAllPhotosFlow() -> AnyPublisher<[Photo], Error>

    // Category Operations
    func getPhotosByCategory(_ categoryId: Int64) async throws -> [Photo]
    func getPhotosByCategoryFlow(_ categoryId: Int64) -> AnyPublisher<[Photo], Error>
    func deletePhotosByCategory(_ categoryId: Int64) async throws
    func getPhotoCategoryCount(_ categoryId: Int64) async throws -> Int
    func getPhotoCountByCategory(categoryId: Int64) async throws -> Int

    // Utility Operations
    func getPhotoCount() async throws -> Int
    func removeFromLibrary(_ photo: Photo) async throws
    func removeFromLibraryById(_ photoId: Int64) async throws
    func cleanupOrphanedPhotos() async throws -> Int

}